File: core/src/commonMain/kotlin/com/encatch/core/MessageProtocol.kt
package com.encatch.core

import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonObject

/** Messages emitted by the hosted form page (WebView) to the native SDK. */
enum class FormMessageType(val wireValue: String) {
    READY("form:ready"),
    SUBMIT("form:submit"),
    COMPLETE("form:complete"),
    CLOSE("form:close"),
    ERROR("form:error"),
    RESIZE("form:resize"),
    LAYOUT("form:layout"),
    CLOSE_BUTTON("form:closeButton"),
    THEME_DATA("form:themeData"),
    REFINE_TEXT_REQUEST("form:refineTextRequest"),
    STARTED("form:started"),
    ANSWERED("form:answered"),
    SECTION_CHANGE("form:section:change"),
    SHOW("form:show"),
    READY_TO_DISMISS("form:readyToDismiss"),
    UPLOAD_FILE_REQUEST("form:uploadFileRequest"),
    QNA_WITH_AI_REQUEST("form:qnaWithAiRequest"),
    REMIND_ME_LATER("form:remindmelater"),
    CTA_TRIGGERED("form:ctaTriggered");

    companion object {
        fun fromWire(value: String): FormMessageType? = entries.find { it.wireValue == value }
    }
}

/** Envelope for a message received from the WebView, parsed via [kotlinx.serialization]. */
@Serializable
data class FormMessage(
    val type: String,
    val formId: String? = null,
    val data: JsonObject? = null,
) {
    val messageType: FormMessageType? get() = FormMessageType.fromWire(type)
}

/** Messages sent from the native SDK to the hosted form page (WebView), via evaluateJavascript. */
enum class SDKMessageType(val wireValue: String) {
    FORM_CONFIG("sdk:formConfig"),
    THEME("sdk:theme"),
    LOCALE("sdk:locale"),
    RESET_DATA("sdk:resetData"),
    PREFILL_RESPONSES("sdk:prefillResponses"),
    REFINE_TEXT_RESPONSE("sdk:refineTextResponse"),
    SUBMIT_PARTIAL_BEFORE_DISMISS("sdk:submitPartialBeforeDismiss"),
    UPLOAD_FILE_RESPONSE("sdk:uploadFileResponse"),
    UPLOAD_FILE_PROGRESS("sdk:uploadFileProgress"),
    QNA_WITH_AI_RESPONSE("sdk:qnaWithAiResponse"),
    QNA_WITH_AI_CHUNK("sdk:qnaWithAiChunk"),
    QNA_WITH_AI_DONE("sdk:qnaWithAiDone"),
}

/** Envelope for a message sent to the WebView; [dataJson] is a pre-serialized JSON object string, or null. */
data class SDKMessage(
    val type: SDKMessageType,
    val dataJson: String? = null,
)

/** Wire-format for a deferred exit_form CTA action, e.g. `redirect_internal`/`redirect_external`/`app_navigate`. */
object CtaAction {
    const val DISMISS = "dismiss"
    const val APP_NAVIGATE = "app_navigate"
    const val REDIRECT_INTERNAL = "redirect_internal"
    const val REDIRECT_EXTERNAL = "redirect_external"
}

File: swift/Sources/Encatch/Core/MessageProtocol.swift
import Foundation

/// Messages emitted by the hosted form page (WebView) to the native SDK.
public enum FormMessageType: String, Sendable, CaseIterable {
    case ready
    case submit
    case complete
    case close
    case error
    case resize
    case layout
    case closeButton
    case themeData
    case refineTextRequest
    case started
    case answered
    case sectionChange
    case show
    case readyToDismiss
    case uploadFileRequest
    case qnaWithAiRequest
    case remindMeLater
    case ctaTriggered

    public var wireValue: String {
        switch self {
        case .ready: return "form:ready"
        case .submit: return "form:submit"
        case .complete: return "form:complete"
        case .close: return "form:close"
        case .error: return "form:error"
        case .resize: return "form:resize"
        case .layout: return "form:layout"
        case .closeButton: return "form:closeButton"
        case .themeData: return "form:themeData"
        case .refineTextRequest: return "form:refineTextRequest"
        case .started: return "form:started"
        case .answered: return "form:answered"
        case .sectionChange: return "form:section:change"
        case .show: return "form:show"
        case .readyToDismiss: return "form:readyToDismiss"
        case .uploadFileRequest: return "form:uploadFileRequest"
        case .qnaWithAiRequest: return "form:qnaWithAiRequest"
        case .remindMeLater: return "form:remindmelater"
        case .ctaTriggered: return "form:ctaTriggered"
        }
    }

    public static func fromWire(_ value: String) -> FormMessageType? {
        allCases.first { $0.wireValue == value }
    }
}

/// Envelope for a message received from the WebView, parsed via `JSONDecoder`.
public struct FormMessage: Codable, Sendable {
    public var type: String
    public var formId: String?
    public var data: JSONValue?

    public var messageType: FormMessageType? { FormMessageType.fromWire(type) }

    public init(type: String, formId: String? = nil, data: JSONValue? = nil) {
        self.type = type
        self.formId = formId
        self.data = data
    }
}

/// Messages sent from the native SDK to the hosted form page (WebView), via evaluateJavaScript.
public enum SDKMessageType: String, Sendable {
    case formConfig
    case theme
    case locale
    case resetData
    case prefillResponses
    case refineTextResponse
    case submitPartialBeforeDismiss
    case uploadFileResponse
    case uploadFileProgress
    case qnaWithAiResponse
    case qnaWithAiChunk
    case qnaWithAiDone

    public var wireValue: String {
        switch self {
        case .formConfig: return "sdk:formConfig"
        case .theme: return "sdk:theme"
        case .locale: return "sdk:locale"
        case .resetData: return "sdk:resetData"
        case .prefillResponses: return "sdk:prefillResponses"
        case .refineTextResponse: return "sdk:refineTextResponse"
        case .submitPartialBeforeDismiss: return "sdk:submitPartialBeforeDismiss"
        case .uploadFileResponse: return "sdk:uploadFileResponse"
        case .uploadFileProgress: return "sdk:uploadFileProgress"
        case .qnaWithAiResponse: return "sdk:qnaWithAiResponse"
        case .qnaWithAiChunk: return "sdk:qnaWithAiChunk"
        case .qnaWithAiDone: return "sdk:qnaWithAiDone"
        }
    }
}

/// Envelope for a message sent to the WebView; `dataJson` is a pre-serialized JSON object string, or nil.
public struct SDKMessage: Sendable {
    public var type: SDKMessageType
    public var dataJson: String?

    public init(type: SDKMessageType, dataJson: String? = nil) {
        self.type = type
        self.dataJson = dataJson
    }
}

/// Wire-format for a deferred exit_form CTA action, e.g. `redirect_internal`/`redirect_external`/`app_navigate`.
public enum CtaAction {
    public static let dismiss = "dismiss"
    public static let appNavigate = "app_navigate"
    public static let redirectInternal = "redirect_internal"
    public static let redirectExternal = "redirect_external"
}
